internal/server/openapi.go

```diff
@@ -178,31 +185,59 @@ func createComponent(schemas openapi3.Schemas, rst reflect.Type) *openapi3.Schem
 	}
 }
 
-func buildProperty(f reflect.StructField, t, parent reflect.Type, parentSchema *openapi3.Schema) *openapi3.SchemaRef {
+type structDescription struct {
+	typ    reflect.Type
+	schema *openapi3.Schema
+	rules  map[string][]ivalidate.ValidationRule
+}
+
+func buildProperty(f reflect.StructField, t reflect.Type, parent structDescription) *openapi3.SchemaRef {
 	if t.Kind() == reflect.Pointer {
-		return buildProperty(f, t.Elem(), parent, parentSchema)
+		t = t.Elem()
 	}
 
 	s := &openapi3.Schema{}
-	setTagInfo(f, t, parent, s, parentSchema)
+	setTagInfo(f, t, parent.typ, s, parent.schema)
 	setTypeInfo(t, s)
 
+	for _, rule := range parent.rules[f.Name] {
+		rule.DescribeSchema(s)
+	}
+	if ivalidate.IsRequired(parent.rules[f.Name]...) {
+		parent.schema.Required = append(parent.schema.Required, getFieldName(f, parent.typ))
+	}
+
 	if s.Type == "array" {
-		s.Items = buildProperty(f, t.Elem(), parent, parentSchema)
+		s.Items = buildProperty(f, t.Elem(), parent)
 	}
 
 	if s.Type == "object" {
 		s.Properties = openapi3.Schemas{}
 
+		// TODO: cleanup
+		reqV := reflect.New(t)
+		req, ok := reqV.Interface().(ivalidate.Request)
```

Comment:
All the API request structs should implement this `validate.Request` interface, so we'll be able to remove the type check and use the passed in struct.